packages/fast-components-react-base/src/select/select.tsx

```diff
@@ -0,0 +1,499 @@
+import Button from "../button";
+
+export interface SelectState {
+    value: string | string[];
+    displayString: string;
+    selectedItems: ListboxItemProps[];
+    isMenuOpen: boolean;
+}
+
+class Select extends Foundation<SelectHandledProps, SelectUnhandledProps, SelectState> {
+    public static displayName: string = "Select";
+
+    public static defaultProps: Partial<SelectProps> = {
+        multiselectable: false,
+        disabled: false,
+        defaultSelection: [],
+        placeholder: "",
+    };
+
+    /**
+     * Handled props instantiation
+     */
+    protected handledProps: HandledProps<SelectHandledProps> = {
+        isMenuOpen: void 0,
+        disabled: void 0,
+        displayStringFormatter: void 0,
+        form: void 0,
+        labelledBy: void 0,
+        multiselectable: void 0,
+        trigger: void 0,
+        required: void 0,
+        managedClasses: void 0,
+        selectedItems: void 0,
+        defaultSelection: void 0,
+        onValueChange: void 0,
+        placeholder: void 0,
+        autoFocus: void 0,
+    };
+
+    private rootElement: React.RefObject<HTMLDivElement> = React.createRef<
+        HTMLDivElement
+    >();
+
+    /**
+     * constructor
+     */
+    constructor(props: SelectProps) {
+        super(props);
+
+        let initialSelection: ListboxItemProps[];
+        if (this.props.selectedItems !== undefined) {
+            initialSelection = Listbox.getListboxItemDataFromIds(
+                this.props.selectedItems,
+                this.props.children
+            );
+        } else {
+            initialSelection = Listbox.getListboxItemDataFromIds(
+                this.props.defaultSelection,
+                this.props.children
+            );
+        }
+
+        if (!this.props.multiselectable && initialSelection.length > 1) {
+            initialSelection = initialSelection.slice(0, 1);
+        }
+
+        this.state = {
+            selectedItems: initialSelection,
+            value: "",
+            displayString: "",
+            isMenuOpen: this.validateMenuState(false),
+        };
+    }
+
+    public componentDidUpdate(prevProps: SelectProps): void {
+        if (prevProps.selectedItems !== this.props.selectedItems) {
+            const newSelection: ListboxItemProps[] = Listbox.getListboxItemDataFromIds(
+                this.props.selectedItems,
+                this.props.children
+            );
+            this.updateSelection(newSelection);
+        }
+        if (prevProps.multiselectable !== this.props.multiselectable) {
+            if (
+                this.props.multiselectable === false &&
+                this.state.selectedItems.length > 1
+            ) {
+                this.updateSelection([this.state.selectedItems[0]]);
+            }
+            this.toggleMenu(this.checkPropsForMenuState());
+        }
+        if (prevProps.isMenuOpen !== this.props.isMenuOpen) {
+            this.toggleMenu(this.checkPropsForMenuState());
```

Comment:
The issue for me here is that it's very hard to discern all the logical paths that might or might not happen when you are firing function multiple times with unrelated criteria. I would prefer something more like:

```
let updatedSelection: ListboxItemProps[];
let updatedMenuVisibility: boolean;

if (prevProps.selectedItems !== this.props.selectedItems) {
    updatedSelection = Listbox.getListboxItemDataFromIds(
        this.props.selectedItems,
        this.props.children
    );
}

if (prevProps.multiselectable !== this.props.multiselectable) {
    // in the case that the multiselect is turned off but there are multiple items selected,
    // choose the first item in the list of selected items
    if (
        this.props.multiselectable === false &&
        this.state.selectedItems.length > 1
    ) {
        updatedSelection = [this.state.selectedItems[0]];
    }

    updatedMenuVisibility = this.checkPropsForMenuState();
}

if (prevProps.isMenuOpen !== this.props.isMenuOpen) {
    updatedMenuVisibility = this.checkPropsForMenuState();
}

if (updatedSelection !== undefined) {
    this.updateSelection(updatedSelection);
}

if (typeof toggledMenu === "boolean") {
    this.toggleMenu(updatedMenuVisibility);
}
```

Reply:
yep